### DISCUSSION TITLE
MedicationStatement Status or Reference first

### DISCUSSION THREAD

**Discussion by Benusheart** [nazrulworld/fhir.resources#98] [Ideas]
Good morning,

I am working with the MedicationStatement Resousce and I am getting the following errors:

When I first want to set the state, like with the other resousces:
```
ms=MedicationStatement.construct()
ms.status="active"
"ValidationError: 1 validation error for MedicationStatement.
__root__
  Expects one of the field values from this list ['medicationCodeableConcept', 'medicationReference']. (type=value_error)" 
```
Then when I try to use the medicationReference first, this error happens:
```
ms=MedicationStatement.construct()
ms.medicationReference={
        "reference": "medication"
    }



"ValidationError": 1 Validation error for MedicationStatement.
__root__ -> status
  none is not an allowed value (type=type_error.none.not_allowed) "
```
My workaround is this:
```
testdict={
    { "status":{ "active",
    "subject":{ 
            "reference": "patient"
        },
    "medicationReference":{
        "reference": "medication"
    }
}
MedicationStatement(**testdict)
```

Maybe it should be possible to specify the status without an error message, like with the other resources.

Could it be, because of the following code in the MedicationStatement Resource:

```
medicationCodeableConcept: fhirtypes.CodeableConceptType = Field(
        None,
        alias="medicationCodeableConcept",
        title="What medication was taken",
        description=(
            "Identifies the medication being administered. This is either a link to"
            " a resource representing the details of the medication or a simple "
            "attribute carrying a code that identifies the medication from a known "
            "list of medications."
        ),
        # if property is element of this resource.
        element_property=True,
        # Choice of Data Types. i.e medication[x]
        one_of_many="medication",
        one_of_many_required=True,
    )

    medicationReference: fhirtypes.ReferenceType = Field(
        None,
        alias="medicationReference",
        title="What medication was taken",
        description=(
            "Identifies the medication being administered. This is either a link to"
            " a resource representing the details of the medication or a simple "
            "attribute carrying a code that identifies the medication from a known "
            "list of medications."
        ),
        # if property is element of this resource.
        element_property=True,
        # Choice of Data Types. i.e medication[x]
        one_of_many="medication",
        one_of_many_required=True,
        # note: Listed Resource Type(s) should be allowed as Reference.
        enum_reference_types=["Medication"],
    )
```
Maybe one_of_many_required should be false, like in Observation:
```
 valueBoolean: bool = Field(
        None,
        alias="valueBoolean",
        title="Actual result",
        description=(
            "The information determined as a result of making the observation, if "
            "the information has a simple value."
        ),
        # if property is element of this resource.
        element_property=True,
        # Choice of Data Types. i.e value[x]
        one_of_many="value",
        one_of_many_required=False,
    )
    valueBoolean__ext: fhirtypes.FHIRPrimitiveExtensionType = Field(
        None, alias="_valueBoolean", title="Extension field for ``valueBoolean``."
    )

    valueCodeableConcept: fhirtypes.CodeableConceptType = Field(
        None,
        alias="valueCodeableConcept",
        title="Actual result",
        description=(
            "The information determined as a result of making the observation, if "
            "the information has a simple value."
        ),
        # if property is element of this resource.
        element_property=True,
        # Choice of Data Types. i.e value[x]
        one_of_many="value",
        one_of_many_required=False,
    )
```
Many greetings
Benusheart

**Comment by nazrulworld** [comment]
Please have a look https://github.com/nazrulworld/fhir.resources/issues/56

**Reply by Benusheart** [reply]
hey :) 

thank you very much! I had read the issue, just seemed to overread point 4, because it works for all other classes so far.  ( e.g. Flag, Observation, Medication, Patient, Questionnaireresponse).
Please don't take this next one as a criticism. I think it's great that you created this library and still maintain it! My concern in this case is that it gets even better. 

EffectivePeriod or effectiveDateTime don't cause any problems in this case either, but these are set to false for required.
Would this be a workaround for now?

Or would warnings be conceivable instead of an exception as soon as a class expects more than the status required?

I don't know your code well enough, I don't know if these are minor or major changes. 
Otherwise it would be great if you don't delete the discussion, if someone else has this problem they can see how to solve it with a dictionary if they can't figure it out themselves.

Thank you for your time!

**Comment by nazrulworld** [comment]
**I will not delete this conversation.** 

**Let's see your point 4**

I will take your example resources MedicationStatement & Observation.
1. Observation: if you look at https://www.hl7.org/fhir/observation-definitions.html#Observation.value_x_  value (valueBoolean, ...., ) is optional (Cardinality 0..1)
2. MedicationStatement:  but if look at https://www.hl7.org/fhir/medicationstatement-definitions.html#MedicationStatement.medication_x_  (medicationX) is required.
3. So for the same reason, you didn’t face any trouble with   effectivePeriod, effectiveDateTime https://www.hl7.org/fhir/medicationstatement-definitions.html#MedicationStatement.effective_x_ is optional

**Reply by Benusheart** [reply]
I definitely understood the difference and found a solution for myself. It's about an idea to avoid this error for other users.

That is why I pointed out the difference. I see three solutions for this problem from the user side:
1. the workaorund
2. set true to false in medication required
3. one changes the construct method in that it throws warnings on multiple requireds if missing and only really needs the status.

The rest I understood, I just want to share an idea to improve or beautify the code
